app/com/linkedin/drelephant/mapreduce/MapReduceFSFetcherHadoop2.java

```diff
@@ -0,0 +1,335 @@
+    logger.info("History done dir: " + _historyLocation);
+  }
+
+  /**
+   * The location of a job history file is in format: {done-dir}/yyyy/mm/dd/{serialPart}.
+   * yyyy/mm/dd is the year, month and date of the finish time.
+   * serialPart is the first 6 digits of the serial number considering it as a 9 digits number.
+   * PS: The serial number is the last part of an app id.
+   * <p>
+   * For example, if appId = application_1461566847127_84624, then serial number is 84624.
+   * Consider it as a 9 digits number, serial number is 000084624. So the corresponding
+   * serialPart is 000084. If this application finish at 2016-5-30, its history file will locate
+   * at {done-dir}/2016/05/30/000084
+   * </p>
+   * <p>
+   * Furthermore, this location format is only satisfied for finished jobs in {done-dir} and not
+   * for running jobs in {intermediate-done-dir}.
+   * </p>
+   */
+  private String getHistoryDir(AnalyticJob job) {
+    // generate the date part
+    Calendar timestamp = Calendar.getInstance();
+    timestamp.setTimeInMillis(job.getFinishTime());
+    String datePart = String.format(TIMESTAMP_DIR_FORMAT,
+            timestamp.get(Calendar.YEAR),
+            timestamp.get(Calendar.MONTH) + 1,
+            timestamp.get(Calendar.DAY_OF_MONTH));
+
+    // generate the serial part
+    String appId = job.getAppId();
+    int serialNumber = Integer.parseInt(appId.substring(appId.lastIndexOf('_') + 1));
+    String serialPart = String.format("%09d", serialNumber)
+            .substring(0, SERIAL_NUMBER_DIRECTORY_DIGITS);
+
+    return _historyLocation + File.separator + datePart + File.separator + serialPart + File.separator;
+  }
+
+  private DataFiles getHistoryFiles(AnalyticJob job) throws IOException {
+    String jobId = Utils.getJobIdFromApplicationId(job.getAppId());
+    String jobConfPath = null;
+    String jobHistPath = null;
+
+    // Search files in done dir
+    String jobHistoryDirPath = getHistoryDir(job);
+    RemoteIterator<LocatedFileStatus> it = _fs.listFiles(new Path(jobHistoryDirPath), false);
+    while (it.hasNext() && (jobConfPath == null || jobHistPath == null)) {
+      String name = it.next().getPath().getName();
+      if (name.contains(jobId)) {
+        if (name.endsWith("_conf.xml")) {
+          jobConfPath = jobHistoryDirPath + name;
+        } else if (name.endsWith(".jhist")) {
+          jobHistPath = jobHistoryDirPath + name;
+        }
+      }
+    }
+
+    // If some files are missing, search in the intermediate-done-dir in case the HistoryServer has
+    // not yet moved them into the done-dir.
+    String intermediateDirPath = _intermediateHistoryLocation + File.separator + job.getUser() + File.separator;
+    if (jobConfPath == null) {
+      jobConfPath = intermediateDirPath + jobId + "_conf.xml";
+      if (!_fs.exists(new Path(jobConfPath))) {
+        throw new FileNotFoundException("Can't find config of " + jobId + " in neither "
```

Comment:
This happens most probably when we look into intermediate-done-dir to find the missing job_conf.xml and at the mean time Job History Server move this file into the done-dir. The fetch will succeed in next retry.

About the frequency, it's arround 13/14482=0.0092%. I count this by parsing one of the log files. In this log 14482 MapReduce apps are analyzed and this happens 13 times.